src/coreclr/src/debug/ee/debuggermodule.cpp

```diff
@@ -239,12 +239,34 @@ void DebuggerModuleTable::RemoveModule(Module* module, AppDomain *pAppDomain)
     }
     CONTRACTL_END;
 
-    // If this is a domain neutral module, then scan the complete list of DebuggerModules looking
-    // for the one with a matching appdomain id.
-    // Note: we have to make sure to lookup the module with the app domain parameter if the module lives in a shared
-    // assembly or the system assembly. <BUGNUM>Bugs 65943 & 81728.</BUGNUM>
-    _ASSERTE( FALSE );
+    LOG((LF_CORDB, LL_INFO1000, "DMT::RM Attempting to remove Module:0x%x AD:0x%x\n", pModule, pAppDomain));
 
+    _ASSERTE(ThreadHoldsLock());
+    _ASSERTE(pModule != NULL);
+
+    HASHFIND hf;
+    DebuggerModuleEntry *pDME; pDME = (DebuggerModuleEntry *) FindFirstEntry(&hf);
+
+    for (pDME = (DebuggerModuleEntry *) FindFirstEntry(&hf);
+         pDME != NULL;
+         pDME = (DebuggerModuleEntry*) FindNextEntry(&hf))
+    {
+        DebuggerModule *pDM = pDME->module;
+        Module* pRtModInDebugMod = pDM->GetRuntimeModule();
+
+        if (pRtModInDebugMod == pModule && pDM->GetAppDomain() == pAppDomain)
+        {
+            LOG((LF_CORDB, LL_INFO1000, "DMT::RM Removing DebuggerMod:0x%x - Module:0x%x DF:0x%x AD:0x%x\n",
+                pDM, pModule, pDM->GetDomainFile(), pAppDomain));
+            TRACE_FREE(pDME->module);
+            DeleteInteropSafe(pDM);
+            Delete(HASH(pRtModInDebugMod), (HASHENTRY *) pDME);
```

Comment:
I assume since a lock held the order of operations here doesn't matter.  

I assume these are roughly
+ Delete the module
+ Remove module entry from the dictionary

Reply:
Doesn't matter mostly because of the lock indeed. The remove from the hash never tries any cleanup, just deletes the entry in the linked list.

As long as no one holds a reference to the debug module the first operation is safe. So far what I've seen is we're good citizens and use LookupOrCreate variants except for FuncEvals, and within the debugger modules themselves. Most cases would run one one thread under a lock, so all is good. The one in the debugger module is the primary module, and that should probably be removed as I couldn't find any reference.